File: 2022/rust/07.rs
//usr/bin/true; rustc -o "/tmp/$0.bin" 1>&2 "$0" && "/tmp/$0.bin" "$@"; exit $?
use std::fs;

fn main() {
    let mut path = Vec::new();
    let mut dirs = std::collections::HashMap::new();

    fs::read_to_string("../inputs/07.txt")
        .unwrap()
        .trim()
        .split("\n")
        .for_each(|l| {
            match l.split(" ").collect::<Vec<&str>>()[..] {
                ["$", "cd", "/"] => {
                    path = Vec::new();
                }
                ["$", "cd", ".."] => {
                    path.pop();
                }
                ["$", "cd", dir] => path.push(dir),
                ["$" | "dir", ..] => (),
                [size, _] => {
                    (0..path.len() + 1).for_each(|i| {
                        *dirs.entry(path[..i].join("/")).or_insert(0) += size
                            .parse::<u64>()
                            .unwrap();
                    });
                }
                _ => (),
            }
        });

    println!(
        "Part 1: {}",
        dirs
            .values()
            .filter(|&&s| s <= 100_000)
            .sum::<u64>()
    );

    println!(
        "Part 2: {}",
        dirs
            .values()
            .filter(|&&s| 70_000_000 - dirs.values().max().unwrap() + s >= 30_000_000)
            .min()
            .unwrap()
    );
}

File: 2020/07.py
import re

rs,bs,os = " bags contain | bags?\.| bags?, ",{},{}

a1,a2 = 0,0
for l in [re.split(rs, l) for l in open("../inputs/2020/07.txt", "r").readlines()]:
    bs[l[0]] = {}
    for b in [b.split(" ", 1) for b in l[1:-1]]:
        if b[0] == "no": continue
        bs[l[0]][b[1]] = int(b[0])
        if b[1] == "shiny gold": os[l[0]] = 1

def func(c):
    l = {}
    for b,v in bs.items():
        if c in v: l = {**{b: 1}, **l, **func(b)}
    return l

def add(c):
    return sum([add(b) * v + v for b,v in bs[c].items()])

print("Part 1:", len(func("shiny gold")))
print("Part 2:", add("shiny gold"))